dnf-docker-test/features/steps/rpm_utils.py

```diff
@@ -110,5 +110,13 @@ def analyze_state(pre, post):
         return State.downgraded
     # In theory, it will never happen as sha1header should be different
     assert ver_cmp == 0
+
+    pre_nevra = hdr2nevra(pre)
+    post_nevra = hdr2nevra(post)
+
+    # probably reinstalled from different repository
```

Comment:
no, when you **re**install it, it should be same package (i.e. header).

Reply:
Well, headers are obviously different in this case:
```
    When I save rpmdb                                                                   # behave/steps/rpm_steps.py:17
    And I enable repository "test4"                                                     # behave/steps/repo_steps.py:488
    And I successfully run "dnf -y repository-packages test2 remove-or-reinstall TestC" # behave/steps/command_steps.py:32
    Then rpmdb changes are                                                              # behave/steps/rpm_steps.py:25
      | State       | Packages |
      | reinstalled | TestC    |
      Assertion Failed: u'TestC-1-1.noarch' -> u'TestC-1-1.noarch'
      shell.command: ['dnf-2', '-y', 'repository-packages', 'test2', 'remove-or-reinstall', 'TestC']
      shell.command.stdout:
      test4                                           758 kB/s | 776  B     00:00    
      Last metadata expiration check: 0:00:00 ago on Wed Oct 18 13:03:04 2017.
      Dependencies resolved.
      =========================================================================

       Package           Arch               Version           Repository         Size
      =========================================================================
      Reinstalling:
       TestC             noarch             1-1               test4             5.6 k
      
      Transaction Summary
      =========================================================================
      
      Total size: 5.6 k
      Downloading Packages:
      Running transaction check
      Transaction check succeeded.
      Running transaction test
      Transaction test succeeded.
      Running transaction
        Preparing        :                                                        1/1 
        Reinstalling     : TestC-1-1.noarch                                       1/2 
        Erasing          : TestC-1-1.noarch                                       2/2 
        Verifying        : TestC-1-1.noarch                                       1/2 
        Verifying        : TestC-1-1.noarch                                       2/2 
      
      Reinstalled:
        TestC.noarch 1-1                                                              
      
      Complete!
```

It fails on: https://github.com/rpm-software-management/ci-dnf-stack/blob/master/dnf-docker-test/features/steps/rpm_utils.py#L114

Reply:
@edynox if headers are different, it is different package. Then reinstall should not reinstal lthis package -> bug in dnf.

Reply:
The package is initialized like this, so headers may differ, but as far as I know, nevra should be a key in DNF, not a header.
```
        Given repository "test2" with packages
           | Package | Tag      | Value |
           | TestB   |          |       |
           | TestC   | Requires | TestD |
        Given repository "test4" with packages
           | Package | Tag      | Value  |
           | TestC   |          |        |
```

Reply:
@edynox so you've found a bug in DNF. now you can fix it ;)

Reply:
I don't think that this is a bug. When a user wants to have some package installed by nevra, he doesn't care much about the header.
Once you run `dnf repository-packages test2 remove-or-reinstall TestC` then you expect that package `TestC-1-1.noarch` will be reinstalled from some other repository (possibly with different signature and header) if available or removed when its not.

Reply:
but then it's not reinstall. it is removing one package and installing another. just with same name/version/release

Reply:
I agree. The wording is misleading. E.g. `remove-or-replace` + `replaced`/`replacing` would be better.